File: orig/hhh-2008-03-29/Sprite.cc
#include "Sprite.h"
#include "Defines.h"
#include <string>

#define delay_set 3

class Sprite;

Sprite::Sprite(char* spritename, AnimationFactory* af){
	//printf(">-------Creating sprite\n");
	this->name = spritename;
	this->af = af;
	this->delay = delay_set;
	this->currentframe = 0;
	string s(spritename);
	s = s + "-default";
	//Animation* def = new Animation(spritename, "default", true, spritename);
	//animations.push_back(def);
	//printf("Adding animation to store\n");
	animations.push_back(af->getByName(s.c_str()));
//	animations.push_back(af->getLast());
	//current = def;
	current = animations[animations.size()-1];
	alive = true;
};

Sprite::~Sprite(){
	//int i;
	//printf("<-------Killing sprite\n");
	//for(i=0; i<animations.size(); i++){
	//	delete(animations[i]);
	//}
};

bool Sprite::animationFinished(){
//	if(!current->loop && current->currentframe < current->numframes){
//		return false;
//	} else if(!current->loop && current->currentframe >= current->numframes){
//		return true;
//	} else {
//		return false;
//	}
	if(current->numframes == currentframe && !current->loop){
		return true;
	} else {
		return false;
	}
};

bool Sprite::isAnimation(char* name){
	if(strcmp(name,current->name)==0){
		return true;
	} else {
		return false;
	}
};

char* Sprite::getAnimation(){
	return current->name;
};
/*
void Sprite::loadAnimation(char* animName, bool loop){
	Animation* tmp = new Animation(this->name,animName,loop);
	animations.push_back(tmp);
};*/

void Sprite::setAnimation(char* name){
	alive = true;
	string s(name);
	string nn(this->name);
	s = nn + "-" + s;
	int cur;
	//printf("Setting animation to %s\n",s.c_str());
	for(cur=0; cur<af->size(); cur++){
		//printf("Current animation is called %s\n",animations[cur]->name);
		Animation* tmp = (Animation*)(af->getByNumber(cur));
		if(strcmp(tmp->getKnown(),s.c_str()) == 0){
			current = tmp;
			currentframe = 0;
		}
	}
};

void Sprite::kill(){
	setAnimation("die");
};

SDL_Surface* Sprite::frame(){
	//setAnimation("default");
	SDL_Surface* tmp = current->getFrame(currentframe);
	delay--;
	if(delay==0){
		currentframe++;
		delay = delay_set;
	}
	if(currentframe>=current->numframes && current->loop){
		currentframe = 0;
	}
	return tmp;
};

void Sprite::move(){
	if(xpos < nextxpos){
		xpos = xpos + speed;
	} else if(xpos > nextxpos){
		xpos = xpos - speed;
	} else if(ypos < nextypos){
		ypos = ypos + speed;
	} else if(ypos > nextypos){
		ypos = ypos - speed;
	}
};

void Sprite::gridx(int x){
	gridxpos = x;
};

void Sprite::gridy(int y){
	gridypos = y;
};

void Sprite::x(int x){
	xpos = x;
};

void Sprite::y(int y){
	ypos = y;
};

int Sprite::x(){
	//xpos = (gridxpos * SIZET) + 50;
	return xpos;
};

int Sprite::y(){
	return ypos;
};

File: sprite.go
package main

import "github.com/qeedquan/go-media/sdl"

type Sprite struct {
	nextxpos           int
	nextypos           int
	direction          int // 1 = up, 2 = down, 3 = left, 4 = right, 5 = stationary
	alive              bool
	animations         []*Animation
	current            *Animation
	af                 *AnimationFactory
	name               string
	xpos, ypos         int
	gridxpos, gridypos int
	currentFrame       int
	speed              int
	delay              int
}

func newSprite(spriteName string, af *AnimationFactory) *Sprite {
	animations := []*Animation{af.getByName(spriteName + "-default")}
	return &Sprite{
		name:       spriteName,
		af:         af,
		delay:      delaySet,
		alive:      true,
		animations: animations,
		current:    animations[0],
	}
}

func (s *Sprite) x() int {
	return s.xpos
}

func (s *Sprite) y() int {
	return s.ypos
}

func (s *Sprite) setX(x int) {
	s.xpos = x
}

func (s *Sprite) setY(y int) {
	s.ypos = y
}

func (s *Sprite) frame() *sdl.Texture {
	frame := s.current.getFrameN(s.currentFrame)
	if s.delay--; s.delay == 0 {
		s.currentFrame++
		s.delay = delaySet
	}
	if s.currentFrame >= len(s.current.frames) && s.current.loop {
		s.currentFrame = 0
	}
	return frame
}

func (s *Sprite) isAnimation(name string) bool {
	return s.current.name == name
}

func (s *Sprite) setAnimation(name string) {
	s.alive = true
	for i := 0; i < af.size(); i++ {
		a := af.getByNumber(i)
		if a.getKnown() == s.name+"-"+name {
			s.current = a
			s.currentFrame = 0
			return
		}
	}
	panic("unreachable")
}

func (s *Sprite) animationFinished() bool {
	if s.currentFrame == len(s.current.frames) && !s.current.loop {
		return true
	}
	return false
}
